### PR TITLE
update vmap to accept nones

### LINKED ISSUE #439: vmap should support out_dims=None
Currently, all the outputs of vmap must be Tensors. We can relax this constraint by letting the user specify `None` for an out_dim (which means, return the object as-is). 

```py
def f(x):
  return 1

x = torch.randn(3)
result = vmap(f, out_dims=(None,))(x)
assert result == 1
```

There are some interesting cases (that we should look to JAX to for what the semantics should be):
- what happens if we specify out_dims=None on a Tensor return, but the Tensor is being vmapped over?

### LINKED ISSUE #1082: vmap should be able to accept `None` in its out_dims argument
Currently, all the outputs of vmap must be Tensors. We can relax this constraint by letting the user specify `None` for an out_dim (which means, return the object as-is). 

```py
def f(x):
  return 1

x = torch.randn(3)
result = vmap(f, out_dims=(None,))(x)
assert result == 1
```

There are some interesting cases (that we should look to JAX to for what the semantics should be):
- what happens if we specify out_dims=None on a Tensor return, but the Tensor is being vmapped over?

### PR DESCRIPTION
**Pull request by srossross**
* Fixes https://github.com/pytorch/functorch/issues/1082 
* Fixes https://github.com/pytorch/functorch/issues/439

### CODE REVIEW

File: torch/_functorch/vmap.py [hunk 1]
```diff
@@ -130,21 +130,30 @@ def _create_batched_inputs(
...
     return tree_unflatten(batched_inputs, args_spec)
 
-# Undos the batching (and any batch dimensions) associated with the `vmap_level`.
+
+def _maybe_remove_batch_dim(name, batched_output, vmap_level, batch_size, out_dim):
+    if out_dim is None:
+        if isinstance(batched_output, torch.Tensor):
+            raise ValueError(
+                f'vmap({name}, ...): `{name}` can not return a '
+                f'Tensor when out_dim is None'
```
Comment (kshitij12345) [inline torch/_functorch/vmap.py:139]: I think the following snippet won't work but it should.

```python
def fn(dummy):
    return torch.randn(3)

vmap(fn, out_dims=(None,))(dummy)
```

ref: https://github.com/pytorch/functorch/issues/439#issuecomment-1029291469

Comment (kshitij12345) [reply torch/_functorch/vmap.py:139]: Also, considering the above case, I think the message should say `can not return a BatchedTensor when out_dim is None`

File: torch/_functorch/vmap.py [hunk 2]
```diff
@@ -130,21 +130,30 @@ def _create_batched_inputs(
...
+            )
+        return batched_output
+
+    if not isinstance(batched_output, torch.Tensor):
+        raise ValueError(f'vmap({name}, ...): `{name}` must only return '
+                         f'Tensors, got type {type(batched_output)} as a return.')
```
Comment (kshitij12345) [inline torch/_functorch/vmap.py:145]: I think the error should be update to mention that you can set `out_dim` to `None` to return a non BatchedTensor as output.

File: torch/_functorch/vmap.py [hunk 3]
```diff
@@ -159,34 +168,39 @@ def incompatible_error():
...
         elif isinstance(out_dims, tuple) and len(out_dims) == 1:
             flat_out_dims = out_dims
-            out_dims = out_dims[0]
+        elif out_dims is None:
+            flat_out_dims = [out_dims]
         else:
             incompatible_error()
     else:
         flat_out_dims = _broadcast_to_and_flatten(out_dims, output_spec)
+        print("flat_out_dims", flat_out_dims)
```
Comment (kshitij12345) [inline torch/_functorch/vmap.py:177]: Stray `print`

File: torch/_functorch/vmap.py [hunk 4]
```diff
@@ -159,34 +168,39 @@ def incompatible_error():
...
         if flat_out_dims is None:
+            print("out_dims, output_spec", out_dims, output_spec)
```
Comment (kshitij12345) [inline torch/_functorch/vmap.py:179]: Stray `print`.

**Review by kshitij12345 — COMMENTED** [review]
I think we need to add support for returning unbatched tensor and a test for the same.

File: torch/_functorch/vmap.py (hunk 4 above)
Comment (srossross) [reply torch/_functorch/vmap.py]: removed

File: torch/_functorch/vmap.py (hunk 2 above)
Comment (srossross) [reply torch/_functorch/vmap.py]: I added `Did you mean to set out_dim= to None for output? `

File: torch/_functorch/vmap.py (hunk 1 above)
Comment (srossross) [reply torch/_functorch/vmap.py]: fixed and added a test case

File: test/functorch/test_vmap.py [hunk 5]
```diff
@@ -317,6 +317,49 @@ def foo(x):
...
         self.assertEqual(result, expected)
 
+    def test_out_dims_none_tuple(self):
+        def foo(x):
+            return x, 'hello world'
+
+        tensor = torch.randn(2, 3)
+        result = vmap(foo, out_dims=(0, None))(tensor)
+        self.assertEqual(result[1], 'hello world')
+        self.assertIsInstance(result[0], torch.Tensor)
```
Comment (kshitij12345) [inline test/functorch/test_vmap.py:327]: Would using `self.assertEqual(result[0], tensor)` make more sense as it is more stricter that `assertIsInstance`?

File: torch/_functorch/vmap.py [hunk 6]
```diff
@@ -130,21 +131,33 @@ def _create_batched_inputs(
...
+            raise ValueError(
+                f'vmap({name}, ...): `{name}` can not return a '
+                f'BatchedTensor when out_dim is None'
+            )
+        return batched_output
+
+    # out_dim is non None
+    if not isinstance(batched_output, torch.Tensor):
+        raise ValueError(f'vmap({name}, ...): `{name}` must only return '
+                         f'Tensors, got type {type(batched_output)} as a return. '
```
Comment (kshitij12345) [inline torch/_functorch/vmap.py:148]: I think `but got type {type(batched_output)}` should be enough i.e. no need for `as a return` (But not a strong opinion)

**Review by kshitij12345 — APPROVED** [review]
LGTM! Thank you @srossross

File: test/functorch/test_vmap.py (hunk 5 above)
Comment (srossross) [reply test/functorch/test_vmap.py:327]: added

File: torch/_functorch/vmap.py (hunk 6 above)
Comment (srossross) [reply torch/_functorch/vmap.py]: removed

**Review by Chillee — APPROVED** [review]
LGTM as well.

**Pull request closed without merging**